### PR TITLE
Replace signup

### PR DESCRIPTION
**Pull request by alxdarksage**
The first of two PRs to use StudyParticipant and remove SignUp. As a result, when onboarding, clients will be able to save any information about a user when signing up (not just data groups, as supported today). This is easy to replace because SignUp is never returned in the API and we pluck the information we need out of the model object to create an account.
- Remove SignUp and substitute StudyParticipant
- Consolidate SignUpValidator and StudyParticipantValidate (latter is a superset of the former already)
- Improved immutability of StudyParticipant so it can be compared to other StudyParticipant objects using equals()

### CODE REVIEW

File: app/org/sagebionetworks/bridge/models/accounts/StudyParticipant.java [hunk 1]
```diff
@@ -112,6 +111,32 @@ public String getId() {
...
     
+    @Override
+    public int hashCode() {
+        return Objects.hash(attributes, consentHistories, createdOn, dataGroups, email, 
+                externalId, firstName, healthCode, id, languages, lastName, notifyByEmail, 
+                password, roles, sharingScope, status);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/models/accounts/StudyParticipant.java:121]: Add equalsVerifier test?

File: app/org/sagebionetworks/bridge/models/accounts/StudyParticipant.java [hunk 2]
```diff
@@ -122,14 +147,33 @@ public String getId() {
         private boolean notifyByEmail;
         private Set<String> dataGroups = ImmutableSet.of();
         private String healthCode;
-        private Map<String,String> attributes = Maps.newHashMap();
-        private Map<String,List<UserConsentHistory>> consentHistories = Maps.newHashMap();
-        private Set<Roles> roles = Sets.newHashSet();
+        private Map<String,String> attributes = ImmutableMap.of();
+        private Map<String,List<UserConsentHistory>> consentHistories = ImmutableMap.of();
+        private Set<Roles> roles = ImmutableSet.of();
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/models/accounts/StudyParticipant.java]: If you want to make these immutable, you should also have the setters or build() make immutable copies.

File: test/org/sagebionetworks/bridge/TestUserAdminHelper.java [hunk 3]
```diff
@@ -46,17 +46,25 @@
...
-            this.roles = Sets.newHashSet(signUp.getRoles());
+        public TestUser(String email, String password, Set<Roles> roles, Set<String> dataGroups, Study study,
+                UserSession session) {
+            this.email = email;
+            this.password = password;
+            this.roles = Sets.newHashSet(TEST_USERS);
+            if (roles != null) {
+                this.roles.addAll(roles);    
+            }
             this.roles.add(TEST_USERS);
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/TestUserAdminHelper.java]: Minor: You create a set with TEST_USERS, add a bunch of roles, and then add TEST_USER.

File: test/org/sagebionetworks/bridge/stormpath/StormpathAccountDaoTest.java [hunk 4]
```diff
@@ -135,13 +136,6 @@ public void getStudyPagedAccounts() {
             accounts = accountDao.getPagedAccountSummaries(study, 0, 20, "bridgeit@");
             assertEquals(1, accounts.getItems().size());
             assertEquals("bridgeit@sagebase.org", accounts.getItems().get(0).getEmail());
-            
-            accounts = accountDao.getPagedAccountSummaries(study, 0, 20, "bridge-testing+SADT");
-            assertTrue(accounts.getItems().size() > 0);
-            for (AccountSummary summary : accounts.getItems()) {
-                assertNull(summary.getFirstName());
-                assertNull(summary.getLastName());
-            }
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/stormpath/StormpathAccountDaoTest.java:144]: Is this deleted test code still needed?

Comment (alxdarksage) [inline test/org/sagebionetworks/bridge/stormpath/StormpathAccountDaoTest.java:144]: Yes. This almost seems like an error I made when merging, this shouldn't be removed. Will restore.

File: app/org/sagebionetworks/bridge/models/accounts/StudyParticipant.java (hunk 1 above)
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/models/accounts/StudyParticipant.java:121]: Added

File: app/org/sagebionetworks/bridge/models/accounts/StudyParticipant.java (hunk 2 above)
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/models/accounts/StudyParticipant.java]: Shuffled this around so collections are copied into immutable collections before constructor is called.

File: test/org/sagebionetworks/bridge/TestUserAdminHelper.java (hunk 3 above)
Comment (alxdarksage) [inline test/org/sagebionetworks/bridge/TestUserAdminHelper.java]: Cleaned up

File: test/org/sagebionetworks/bridge/stormpath/StormpathAccountDaoTest.java (hunk 4 above)
Comment (alxdarksage) [inline test/org/sagebionetworks/bridge/stormpath/StormpathAccountDaoTest.java:144]: Added this back.

**Pull request merged**